File: PhaseGate/CheckOut/check_out.py
import datetime

cart = []

def add_item_to_cart(cart: list) -> str:
    item_name = input("What did user buy? ")
    item_quantity = int(input("How many pieces: "))
    item_price = float(input("How much per item: "))
    cart.append(dict(
        item_name=item_name,
        item_quantity=item_quantity,
        item_price=item_price,
        total=item_price * item_quantity
    ))
    return f"Item {item_name} added to cart."


def checkout_invoice(customer: str, cashier: str, discount: float, cart: list) -> str:
    subtotal = sum(item['total'] for item in cart)
    vat = subtotal * 0.075
    billing = subtotal + vat - discount
    invoice = "\n".join(
        f"{item['item_name']}\t {item['item_quantity']}\t ₦{item['item_price']:.2f}\t ₦{item['total']:.2f}"
        for item in cart
    )
    return (f"""
    SEMICOLON STORES
    MAIN BRANCH
    LOCATION: 312, HERBERT MACAULAY WAY, SABO YABA, LAGOS
    TEL: 080123456789
    Date: {datetime.datetime.now().strftime("%a' %d %b %Y %H:%M:%S")}
    Cashier: {cashier}
    Customer Name: {customer}
    =====================================================
    ITEM    QTY     PRICE     TOTAL(NGN)
    -----------------------------------------------------
    {invoice}
    -----------------------------------------------------
                                Sub Total: ₦{subtotal:.2f}
                                Discount: ₦{discount:.2f}
                                VAT: ₦{vat:.2f}
    =====================================================
                                Billing Total: ₦{billing:.2f}
    =====================================================
    THIS IS NOT A RECEIPT KINDLY PAY ₦{billing:.2f}
    =====================================================
    """, billing)


customer_name = input("What is the name of the customer: ")


def process_customer_transaction():
    while True:
        add_item_to_cart(cart)
        more_item_option = input("Do you want to add more items? (y/n): ")
        if more_item_option.lower() in ["y", "yes", "ok"]:
            continue
        cashier = input("What is your name? ")
        discount = float(input("How much discount would the customer get? "))
        invoice = checkout_invoice(customer_name, cashier, discount, cart)[0]
        billing = checkout_invoice(customer_name, cashier, discount, cart)[1]
        print(invoice)
        while True:
            customer_paid = float(input("How much did customer pay? "))
            if customer_paid == billing:
                print(f"{customer_name} paid ₦{customer_paid:.2f} successfully!")
                break
            elif customer_paid > billing:
                print(f"{customer_name} has paid but would be balanced with ₦{customer_paid-billing:.2f}")
                break
            elif customer_paid < billing:
                print(f"{customer_name}'s payment is incomplete remaining ₦{billing - customer_paid:.2f} to complete payment.")
                continue
        print("Closing transaction for next customer")
        break


while True:
    option = input("Enter 'start' to process customer transaction or (end/exit) to close:")
    if option.lower() in ["exit", "close", "end"]:
        break
    elif option.lower() == "start":
        process_customer_transaction()
    else:
        print("Invalid option enter (start or end/exit) to close transaction")
        continue

File: PhaseGate/CheckOut/check_out.js
const prompt = require("prompt-sync")();

cart = [];


const addItemToCart = (cart) => {
    let itemName = prompt("What did user buy? ");
    let itemQuantity = prompt("How many pieces: ");
    let itemPrice = prompt("How much per item: ");
    cart.push({
        itemName:itemName,
        itemQuantity:itemQuantity,
        itemPrice:itemPrice,
        total:itemPrice*itemQuantity
    });
    return `Item ${itemName} added to cart.`
};


const checkoutInvoice = (customer, cashier, discount, cart) => {
    let subtotal = 0;
    cart.forEach(item => {
        subtotal += item.total;
    });
    const vat = subtotal * 0.075;
    const billing = subtotal + vat - discount;

    const invoiceLines = cart.map(item =>
        `${item.itemName}\t ${item.itemQuantity}\t ₦${item.itemPrice}\t ₦${item.total}`
    ).join("\n\t");

    return [`
    SEMICOLON STORES
    MAIN BRANCH
    LOCATION: 312, HERBERT MACAULAY WAY, SABO YABA, LAGOS
    TEL: 080123456789
    Date: ${new Date().toLocaleString()}
    Cashier: ${cashier}
    Customer Name: ${customer}
    =====================================================
        ITEM    QTY     PRICE     TOTAL(NGN)
    -----------------------------------------------------
    ${invoiceLines}
    -----------------------------------------------------
                                Sub Total: ₦${subtotal}
                                Discount: ₦${discount}
                                VAT: ₦${vat}
    =====================================================
                                Billing Total: ₦${billing}
    =====================================================
    THIS IS NOT A RECEIPT KINDLY PAY ₦${billing}
    =====================================================
    `, billing];
};


const processCustomerTransaction = (customerName) => {
    while (true) {
        addItemToCart(cart);
        const moreItemOption = prompt("Do you want to add more items? (y/n): ");

        if (['y', 'yes', 'ok'].includes(moreItemOption.toLowerCase())) {
            continue;
        }

        const cashier = prompt("What is your name? ");
        const discount = parseFloat(prompt("How much discount would the customer get? "));
        const [invoice, billing] = checkoutInvoice(customerName, cashier, discount, cart);

        console.log(invoice);

        while (true) {
            const customerPaid = parseFloat(prompt("How much did customer pay? "));
            if (customerPaid === billing) {
                console.log(`${customerName} paid ₦${customerPaid} successfully!`);
                break;
            } else if (customerPaid > billing) {
                console.log(`${customerName} has paid but would be balanced with ₦${customerPaid - billing}`);
                break;
            } else if (customerPaid < billing) {
                let newBill = billing;
                newBill -= customerPaid;
                console.log(`${customerName}'s payment is incomplete, remaining ₦${newBill} to complete payment.`);
                continue;
            }
        }

        console.log("Closing transaction for next customer");
        break;
    }
};


while (true) {
    const option = prompt("Enter 'start' to process customer transaction or (end/exit) to close:");
    if (['exit', 'close', 'end'].includes(option.toLowerCase())) {
        cart.splice(0, cart.length);
        break;
    } else if (option.toLowerCase() === 'start') {
        const customerName = prompt("What is the name of the customer: ");
        processCustomerTransaction(customerName);
    } else {
        console.log("Invalid option, enter (start or end/exit) to close transaction");
        continue;
    }
}
